ctk/common/src/main/kotlin/org/codice/compliance/verification/core/CommonDataTypeVerifier.kt

```diff
@@ -31,156 +30,170 @@ import java.net.URI
-    }
-    return SplitString(dateTime.substring(0, hyphenIndex), dateTime.substring(hyphenIndex + 1))
-}
-
-// https://www.w3.org/TR/xmlschema-2/#dateTime
-@Suppress("SpreadOperator" /* Judicious use of spread operator to reduce cognitive complexity */)
-private fun verifyYear(year: String, errorCode: SAMLSpecRefMessage?) {
-    // remove the negative sign to make verification easier
-    val strippedYear: String = if (year.indexOf('-') == 0) year.substring(1) else year
-
-    val codes = if (errorCode == null) emptyArray()
-    else arrayOf(errorCode)
-
-    // check if year is an integer && "a plus sign is not permitted"
-    if (!strippedYear.matches("""\d+""".toRegex())) {
-        throw SAMLComplianceException.create(*codes,
-                SAMLCore_1_3_3,
-                XMLDatatypesSchema_3_2_7_1_c,
-                message = "A '+' was found.")
-    }
-
-    // "if more than four digits, leading zeros are prohibited"
-    if (strippedYear.length > FOUR_DIGIT_YEAR_LEN && strippedYear.startsWith('0')) {
-        throw SAMLComplianceException.create(*codes,
-                SAMLCore_1_3_3,
-                XMLDatatypesSchema_3_2_7_1_a,
-                message = "The year value of $strippedYear is invalid.")
-    }
-
-    // "'0000' is prohibited"
-    if (strippedYear == "0000") {
-        throw SAMLComplianceException.create(*codes,
-                SAMLCore_1_3_3,
-                XMLDatatypesSchema_3_2_7_1_b,
-                message = "The year value of $strippedYear is invalid.")
-    }
-}
-
-// todo allow an unlimited amount of fractional seconds as stated in the XML Datatypes Schema 3.2.7
-// todo "SAML system entities SHOULD NOT rely on time resolution finer than milliseconds" Core.1.3.3
-// helper for verifyTimeValues
-private fun verifyRestOfDateTime(restOfDateTime: String, errorCode: SAMLSpecRefMessage?) {
-    val format = DateTimeFormatter.ofPattern("MM'-'dd'T'HH':'mm':'ss['.'SSS]['Z']")
-
-    try {
-        format.parse(restOfDateTime)
-    } catch (e: DateTimeParseException) {
-        if (errorCode != null) throw SAMLComplianceException.create(errorCode,
-                SAMLCore_1_3_3,
-                XMLDatatypesSchema_3_2_7,
-                message = "The date value of $restOfDateTime is incorrectly formatted.",
-                cause = e)
-        else throw SAMLComplianceException.create(SAMLCore_1_3_3,
-                XMLDatatypesSchema_3_2_7,
-                message = "The date value of $restOfDateTime is incorrectly formatted.",
-                cause = e)
+class CommonDataTypeVerifier {
+    companion object {
+        private const val FOUR_DIGIT_YEAR_LEN = 4
+        private const val NO_HYPHEN_MESSAGE = "No hyphen was found in the date value of"
+
+        var ids = mutableListOf<String>()
+
+        /** 1.3 Common Data Types **/
+        fun verifyCommonDataType(samlDom: Node) {
+            ids = mutableListOf()
+            var i = samlDom.childNodes.length - 1
+
+            while (i >= 0) {
```

Comment:
✏️ A ranged loop would be cleaner:
`for (i in samlDom.childNodes.length - 1 downTo 0) {`
you can remove the var and get rid of the decrement on line 59 with that.